### DISCUSSION TITLE
How to check Point In Polygon in GJK?

### DISCUSSION THREAD

**Discussion by IPECTER** [dyn4j/dyn4j#296] [Q&A] [answered]
I want to check Point In Polygon (Point and Polygon) using GJK algorithm..
but GJK class support only Polygon and Polygon check...

**Comment by wnbittle** [comment]
If you are asking how to do this with dyn4j, then I think what you are looking for are the `contains` methods.  For example:

```java
// assume you have the body
Body b = ...;
Convex c = b.getFixture(0);
Vector2 point = new Vector2(1, 2);
if (c.contains(point, b.getTransform(), true)) {
  // the point is contained in the Convex Shape
}
```

For a polygon, there's no need for something as complicated as GJK for point in polygon tests.  [dyn4j uses a O(n) worst case algorithm for point in polygon checks](https://github.com/dyn4j/dyn4j/blob/38a46852affd0306b1bbe347b65afb811173999e/src/main/java/org/dyn4j/geometry/Polygon.java#L316).  It's very simple and efficient because a critical assumption - that the Polygon is Convex.

The process goes like this:

- choose an edge of the Polygon to start with
- determine if the point is on the right or left side of the edge
- for each remaining edge of the Polygon
  - is the point on the same side as found at the beginning?
  - if not, then the point is NOT contained in the Polygon
- if the point is on the same side of every edge of the Polygon, then the point is contained in the Polygon

**Reply by IPECTER** [reply]
Thx!
I have more one question

Im developing check Point In Polygon for simple polygon.
1-1. When program starts, do Simple Polygon Simplification using Ear Clipping.
1-2. get convex polygon max/min x and y.
2-1. When check PIP, check point x and y is in max/min x and y square.
2-2. if true, check Convex#contains(Vector2. Transform, boolean)
is it good?

**Reply by IPECTER** [reply]
"[dyn4j uses a O(n) worst case algorithm for point in polygon checks](https://github.com/dyn4j/dyn4j/blob/38a46852affd0306b1bbe347b65afb811173999e/src/main/java/org/dyn4j/geometry/Polygon.java#L316)."?
Do you mean Convex#contains is slow and worst case algorithm?
Do not use Convex#contains?
make some code following "The process goes like this:"?

GJK polygon and polygon VS Convex#contains point in polygon.
What is faster?

**Reply by wnbittle** [reply, marked as answer]
> Im developing check Point In Polygon for simple polygon.
> 1-1. When program starts, do Simple Polygon Simplification using Ear Clipping.
> 1-2. get convex polygon max/min x and y.
> 2-1. When check PIP, check point x and y is in max/min x and y square.
> 2-2. if true, check Convex#contains(Vector2. Transform, boolean)
> is it good?

Yes, that's a good approach.  There's a `Body.getFixtures(Vector2)` method, but it will be inefficient if the body has 100s of `Convex` shapes.  If you are adding the `Polygon`s to a `Body` and then adding that body to a `World`, it would be a little more efficient to use the `detect` method that accepts an `AABB` for example:
```java
Vector2 simplePolygon = ...;
Decomposer decomposer = new EarClipping();
List<Convex> polygons = decomposer.decompse(simplePolygon);

Body body = new Body();
for (Convex convex : polygons) {
  body.addFixture(convex);
}
body.setMass(MassType.NORMAL);

World world = new World();
world.addBody(body);

Vector2 point = new Vector2(1, 1);

// expand the point a bit
double expansion = 0.1;
AABB aabb = new AABB(point, expansion);
DetectFilter filter = new DetectFilter(true, true, null);
List<DetectResult<Body, BodyFixture>> results = world.detect(aabb, filter);

// process the results
for (DetectResult<Body, BodyFixture> result : results) {
  if (result.getFixture().getShape().contains(point)) {
    // do what you need
  }
}
```
This is more efficient because it will use the `AABB`'s that exist in the `BroadphaseDetector` to quickly find the `BodyFixture`s that overlap.  There's an example of this in the samples [here](https://github.com/dyn4j/dyn4j-samples/blob/747a479d3a0059f73a4de62d33b6003d387d933e/src/main/java/org/dyn4j/samples/framework/input/MousePickingInputHandler.java#L139) which is used to for mouse picking.


> Do you mean Convex#contains is slow and worst case algorithm?

No.  The `Convex.contains(...)` method is optimal.  In the case the point is contained within the Polygon, it's linear time complexity O(n) where n is the number of edges in the Polygon.  In the case the point is NOT contained within the Polygon, it'll be better than O(n) because it will stop as soon as it detects it's not contained.

**Reply by IPECTER** [reply]
Thx for answer!!!
